# Median Income and AMI Dataset Cleaning

Now we will explore the median income and ami dataset so that we can better establish socioeconomic status later on.

In [1]:
# we first need to import 
import pandas as pd
import numpy as np
import os

We will begin by reading in the data.

In [4]:
# get the median income and ami csv file
df = pd.read_csv("Median_Income_and_AMI.csv")
df

tract  med_hh_income  med_hh_income_universe  \
0     6037199700        38892.0                    1204   
1     6037199801        41027.0                     903   
2     6037199802        42500.0                     612   
3     6037199900        37232.0                     845   
4     6037201110        65000.0                     782   
...          ...            ...                     ...   
2490  6037554516       126450.0                    1215   
2491  6037554517       107672.0                    1352   
2492  6037554518       104439.0                    1558   
2493  6037554519       131012.0                    1216   
2494  6037554521        90280.0                    1863   

               ami_category below_med_income below_60pct_med_income  \
0           Very Low Income              Yes                    Yes   
1           Very Low Income              Yes                    Yes   
2           Very Low Income              Yes                    Yes   
3           Very Low Income              Yes                    Yes   
4                Low Income              Yes                     No   
...                     ...              ...                    ...   
2490  Above Moderate Income               No                     No   
2491  Above Moderate Income               No                     No   
2492  Above Moderate Income               No                     No   
2493  Above Moderate Income               No                     No   
2494             Low Income               No                     No   

     below_moderate_income    sup_dist                               csa  \
0                      Yes  District 1  Los Angeles - Wholesale District   
1                      Yes  District 1     Los Angeles - Lincoln Heights   
2                      Yes  District 1     Los Angeles - Lincoln Heights   
3                      Yes  District 1     Los Angeles - Lincoln Heights   
4                      Yes  District 1           Los Angeles - El Sereno   
...                    ...         ...                               ...   
2490                    No  District 4                  City of Cerritos   
2491                    No  District 4                  City of Cerritos   
2492                    No  District 4                  City of Cerritos   
2493                    No  District 4                  City of Cerritos   
2494                   Yes  District 4                  City of Cerritos   

                spa  ESRI_OID   Shape__Area  Shape__Length  
0     SPA 4 - Metro      2347  1.041050e+07   13808.463240  
1     SPA 4 - Metro      2348  3.724107e+06    9459.391827  
2     SPA 4 - Metro      2349  3.296129e+06    8868.744225  
3     SPA 4 - Metro      2350  4.782361e+06   10141.728020  
4     SPA 4 - Metro      2351  1.099246e+07   15893.383640  
...             ...       ...           ...            ...  
2490   SPA 7 - East      4837  1.555650e+07   21274.227410  
2491   SPA 7 - East      4838  1.421767e+07   15905.089170  
2492   SPA 7 - East      4839  1.938903e+07   21218.412990  
2493   SPA 7 - East      4840  1.866694e+07   19500.866810  
2494   SPA 7 - East      4841  5.352220e+07   37625.839750  

[2495 rows x 13 columns]

### Median Income and AMI

We know that the median income and AMI dataset has statistics for median household income and the corresponding AMI category, but what else does it include? Let's take a look!

In [6]:
df.head()

tract  med_hh_income  med_hh_income_universe     ami_category  \
0  6037199700        38892.0                    1204  Very Low Income   
1  6037199801        41027.0                     903  Very Low Income   
2  6037199802        42500.0                     612  Very Low Income   
3  6037199900        37232.0                     845  Very Low Income   
4  6037201110        65000.0                     782       Low Income   

  below_med_income below_60pct_med_income below_moderate_income    sup_dist  \
0              Yes                    Yes                   Yes  District 1   
1              Yes                    Yes                   Yes  District 1   
2              Yes                    Yes                   Yes  District 1   
3              Yes                    Yes                   Yes  District 1   
4              Yes                     No                   Yes  District 1   

                                csa            spa  ESRI_OID   Shape__Area  \
0  Los Angeles - Wholesale District  SPA 4 - Metro      2347  1.041050e+07   
1     Los Angeles - Lincoln Heights  SPA 4 - Metro      2348  3.724107e+06   
2     Los Angeles - Lincoln Heights  SPA 4 - Metro      2349  3.296129e+06   
3     Los Angeles - Lincoln Heights  SPA 4 - Metro      2350  4.782361e+06   
4           Los Angeles - El Sereno  SPA 4 - Metro      2351  1.099246e+07   

   Shape__Length  
0   13808.463240  
1    9459.391827  
2    8868.744225  
3   10141.728020  
4   15893.383640

What about the size of the dataset?

In [7]:
df.shape

(2495, 13)

This means that there are currently 2495 observations in our data, with 13 different categories. What can these categories tell us?

In [13]:
df.columns

Index(['tract', 'med_hh_income', 'med_hh_income_universe', 'ami_category',
       'below_med_income', 'below_60pct_med_income', 'below_moderate_income',
       'sup_dist', 'csa', 'spa', 'ESRI_OID', 'Shape__Area', 'Shape__Length'],
      dtype='object')

In [14]:
df.describe()

tract  med_hh_income  med_hh_income_universe     ESRI_OID  \
count  2.495000e+03    2458.000000             2495.000000  2495.000000   
mean   6.037403e+09   76849.334418             1335.672946  3594.000000   
std    2.303896e+05   35546.132788              533.910989   720.388784   
min    6.037101e+09    4918.000000                0.000000  2347.000000   
25%    6.037209e+09   51157.500000              988.000000  2970.500000   
50%    6.037403e+09   69698.000000             1282.000000  3594.000000   
75%    6.037552e+09   94515.500000             1625.000000  4217.500000   
max    6.037980e+09  250001.000000             5617.000000  4841.000000   

        Shape__Area  Shape__Length  
count  2.495000e+03    2495.000000  
mean   4.579716e+07   20273.817017  
std    4.270580e+08   34074.651233  
min    4.836532e+05    2815.257443  
25%    5.921008e+06   10607.040730  
50%    1.025761e+07   14365.057030  
75%    1.835792e+07   19904.426710  
max    1.608691e+10  915242.577100

Looks like more than half of the columns are categorical and not numerical. What are the other types of data?

In [15]:
# finding types of each column
df.dtypes

tract                       int64
med_hh_income             float64
med_hh_income_universe      int64
ami_category               object
below_med_income           object
below_60pct_med_income     object
below_moderate_income      object
sup_dist                   object
csa                        object
spa                        object
ESRI_OID                    int64
Shape__Area               float64
Shape__Length             float64
dtype: object

In [17]:
df.index

RangeIndex(start=0, stop=2495, step=1)

Something that will be important for our analysis later on is identifying median income for specific regions. Let's take a look at the different locations in this dataset:

In [20]:
df['csa'].unique()

array(['Wholesale District', 'Lincoln Heights', 'El Sereno',
       'Highland Park', 'University Hills', 'Boyle Heights',
       'City of Pomona', 'City of Diamond Bar', 'Rowland Heights',
       'City of Industry', 'City of Walnut', 'Covina', 'City of Covina',
       'City of San Dimas', 'Covina (Charter Oak)', 'City of Glendora',
       'Azusa', 'City of Azusa', 'City of Norwalk', 'City of Artesia',
       'City of Lakewood', 'City of Hawaiian Gardens',
       'City of Long Beach', 'City of Cerritos', 'Little Tokyo',
       'Chinatown', 'Downtown', 'Temple-Beaudry', 'Historic Filipinotown',
       'Westlake', 'City of Irwindale', 'City of Baldwin Park',
       'City of West Covina', 'City of Signal Hill', 'Pico-Union',
       'Hancock Park', 'Wilshire Center', 'Little Bangladesh',
       'Koreatown', 'Bassett', 'City of La Puente', 'West Puente Valley',
       'Valinda', 'San Jose Hills', 'Avocado Heights', 'North Whittier',
       'Hacienda Heights', 'Tujunga', 'Sun Valley', 'Sunlan

Since all of our data is from Los Angeles, we can just look at the city instead.

In [22]:
# function to single out the city/community from the string
def a (city):
    if ' - ' in city:
        idx = city.index('- ')
    else:
        idx = -2
    return city[idx+2:]
df['csa'] = df['csa'].apply(a)

In [23]:
df.head()

tract  med_hh_income  med_hh_income_universe     ami_category  \
0  6037199700        38892.0                    1204  Very Low Income   
1  6037199801        41027.0                     903  Very Low Income   
2  6037199802        42500.0                     612  Very Low Income   
3  6037199900        37232.0                     845  Very Low Income   
4  6037201110        65000.0                     782       Low Income   

  below_med_income below_60pct_med_income below_moderate_income    sup_dist  \
0              Yes                    Yes                   Yes  District 1   
1              Yes                    Yes                   Yes  District 1   
2              Yes                    Yes                   Yes  District 1   
3              Yes                    Yes                   Yes  District 1   
4              Yes                     No                   Yes  District 1   

                  csa            spa  ESRI_OID   Shape__Area  Shape__Length  
0  Wholesale District  SPA 4 - Metro      2347  1.041050e+07   13808.463240  
1     Lincoln Heights  SPA 4 - Metro      2348  3.724107e+06    9459.391827  
2     Lincoln Heights  SPA 4 - Metro      2349  3.296129e+06    8868.744225  
3     Lincoln Heights  SPA 4 - Metro      2350  4.782361e+06   10141.728020  
4           El Sereno  SPA 4 - Metro      2351  1.099246e+07   15893.383640

We should also check for null values.

In [27]:
# checking if csa column has nans
df['csa'].hasnans

False

In [28]:
# checking if ami category colunn has nans
df['ami_category'].hasnans

True

In [29]:
# checking if median_hh_income has nans
df['med_hh_income'].hasnans

True

In [30]:
# finding all the rows in the dataframe where med_hh_income column ins NaN
df.loc[ df['med_hh_income'].isna() ]

tract  med_hh_income  med_hh_income_universe ami_category  \
446   6037216301            NaN                     614          NaN   
502   6037578100            NaN                       0          NaN   
504   6037599100            NaN                      79          NaN   
761   6037222700            NaN                     108          NaN   
896   6037115103            NaN                      11          NaN   
1458  6037265301            NaN                       0          NaN   
1560  6037901003            NaN                       0          NaN   
1658  6037273403            NaN                    1020          NaN   
2052  6037920200            NaN                       0          NaN   
2072  6037980001            NaN                       0          NaN   
2073  6037980002            NaN                       0          NaN   
2074  6037980003            NaN                       0          NaN   
2075  6037980004            NaN                       0          NaN   
2076  6037980005            NaN                       0          NaN   
2077  6037980006            NaN                       0          NaN   
2078  6037980007            NaN                       0          NaN   
2079  6037980008            NaN                      17          NaN   
2080  6037980009            NaN                       0          NaN   
2083  6037980012            NaN                       0          NaN   
2084  6037980013            NaN                       0          NaN   
2085  6037980014            NaN                      11          NaN   
2089  6037980018            NaN                       0          NaN   
2091  6037980020            NaN                       0          NaN   
2092  6037980021            NaN                      14          NaN   
2093  6037980022            NaN                       0          NaN   
2223  6037980023            NaN                       0          NaN   
2225  6037980026            NaN                       0          NaN   
2226  6037980028            NaN                       0          NaN   
2227  6037980030            NaN                       0          NaN   
2228  6037980031            NaN                      22          NaN   
2229  6037980033            NaN                      14          NaN   
2230  6037980034            NaN                       9          NaN   
2232  6037980036            NaN                     164          NaN   
2233  6037980037            NaN                       0          NaN   
2234  6037980038            NaN                       0          NaN   
2235  6037980039            NaN                       0          NaN   
2442  6037402404            NaN                       4          NaN   

     below_med_income below_60pct_med_income below_moderate_income  \
446               NaN                    NaN                   NaN   
502               NaN                    NaN                   NaN   
504               NaN                    NaN                   NaN   
761               NaN                    NaN                   NaN   
896               NaN                    NaN                   NaN   
1458              NaN                    NaN                   NaN   
1560              NaN                    NaN                   NaN   
1658              NaN                    NaN                   NaN   
2052              NaN                    NaN                   NaN   
2072              NaN                    NaN                   NaN   
2073              NaN                    NaN                   NaN   
2074              NaN                    NaN                   NaN   
2075              NaN                    NaN                   NaN   
2076              NaN                    NaN                   NaN   
2077              NaN                    NaN                   NaN   
2078              NaN                    NaN                   NaN   
2079              NaN                    NaN                   NaN   
2080          

In [32]:
# all of the rows having NaN in med_hh_income column have NaNs in ami_catgory, 
# below_med_income, below_60pct_med_income, and below_moderate_income as well
df.loc[ df['med_hh_income'].isna()].shape
df.shape

(2495, 13)

In [33]:
# shape of df once dropping rows with NaN
df.dropna().shape

(2458, 13)

In [34]:
# dropping rows with nans
df.dropna(inplace = True)

In [35]:
df['med_hh_income'].hasnans

False

In [36]:
df['ami_category'].hasnans

False

To optimize run time even more, we can eliminate redundant information. For example, in the csa column, many of the locations are described as 'City of ...'. We will therefore remove this description and only keep the name of the city instead.

In [43]:
df['csa'].unique()

array(['Wholesale District', 'Lincoln Heights', 'El Sereno',
       'Highland Park', 'University Hills', 'Boyle Heights',
       'City of Pomona', 'City of Diamond Bar', 'Rowland Heights',
       'City of Industry', 'City of Walnut', 'Covina', 'City of Covina',
       'City of San Dimas', 'Covina (Charter Oak)', 'City of Glendora',
       'Azusa', 'City of Azusa', 'City of Norwalk', 'City of Artesia',
       'City of Lakewood', 'City of Hawaiian Gardens',
       'City of Long Beach', 'City of Cerritos', 'Little Tokyo',
       'Chinatown', 'Downtown', 'Temple-Beaudry', 'Historic Filipinotown',
       'Westlake', 'City of Irwindale', 'City of Baldwin Park',
       'City of West Covina', 'City of Signal Hill', 'Pico-Union',
       'Hancock Park', 'Wilshire Center', 'Little Bangladesh',
       'Koreatown', 'Bassett', 'City of La Puente', 'West Puente Valley',
       'Valinda', 'San Jose Hills', 'Avocado Heights', 'North Whittier',
       'Hacienda Heights', 'Tujunga', 'Sun Valley', 'Sunlan

In [44]:
# drop the 'City of' part of the string, as that is unnecessary
fncof = lambda x : x if 'City of ' not in x else x[8:]
df['csa'] = df['csa'].apply(fncof)

In [45]:
#checking the unique values
df['csa'].unique()

array(['Wholesale District', 'Lincoln Heights', 'El Sereno',
       'Highland Park', 'University Hills', 'Boyle Heights', 'Pomona',
       'Diamond Bar', 'Rowland Heights', 'Industry', 'Walnut', 'Covina',
       'San Dimas', 'Covina (Charter Oak)', 'Glendora', 'Azusa',
       'Norwalk', 'Artesia', 'Lakewood', 'Hawaiian Gardens', 'Long Beach',
       'Cerritos', 'Little Tokyo', 'Chinatown', 'Downtown',
       'Temple-Beaudry', 'Historic Filipinotown', 'Westlake', 'Irwindale',
       'Baldwin Park', 'West Covina', 'Signal Hill', 'Pico-Union',
       'Hancock Park', 'Wilshire Center', 'Little Bangladesh',
       'Koreatown', 'Bassett', 'La Puente', 'West Puente Valley',
       'Valinda', 'San Jose Hills', 'Avocado Heights', 'North Whittier',
       'Hacienda Heights', 'Tujunga', 'Sun Valley', 'Sunland',
       'Lakeview Terrace', 'Shadow Hills', 'Pacoima', 'Country Club Park',
       'Melrose', 'Park La Brea', 'Carthay', 'Miracle Mile',
       'South Carthay', 'Crestview', 'Mid-city', 'La

Lastly, we can delete unnecessary columns.

In [49]:
df = df.drop(columns=['tract', 'ESRI_OID'])
df.head()

med_hh_income  med_hh_income_universe     ami_category below_med_income  \
0        38892.0                    1204  Very Low Income              Yes   
1        41027.0                     903  Very Low Income              Yes   
2        42500.0                     612  Very Low Income              Yes   
3        37232.0                     845  Very Low Income              Yes   
4        65000.0                     782       Low Income              Yes   

  below_60pct_med_income below_moderate_income    sup_dist  \
0                    Yes                   Yes  District 1   
1                    Yes                   Yes  District 1   
2                    Yes                   Yes  District 1   
3                    Yes                   Yes  District 1   
4                     No                   Yes  District 1   

                  csa            spa   Shape__Area  Shape__Length  
0  Wholesale District  SPA 4 - Metro  1.041050e+07   13808.463240  
1     Lincoln Heights  SPA 4 - Metro  3.724107e+06    9459.391827  
2     Lincoln Heights  SPA 4 - Metro  3.296129e+06    8868.744225  
3     Lincoln Heights  SPA 4 - Metro  4.782361e+06   10141.728020  
4           El Sereno  SPA 4 - Metro  1.099246e+07   15893.383640

Let's take a final look at our dataset!

In [50]:
df

med_hh_income  med_hh_income_universe           ami_category  \
0           38892.0                    1204        Very Low Income   
1           41027.0                     903        Very Low Income   
2           42500.0                     612        Very Low Income   
3           37232.0                     845        Very Low Income   
4           65000.0                     782             Low Income   
...             ...                     ...                    ...   
2490       126450.0                    1215  Above Moderate Income   
2491       107672.0                    1352  Above Moderate Income   
2492       104439.0                    1558  Above Moderate Income   
2493       131012.0                    1216  Above Moderate Income   
2494        90280.0                    1863             Low Income   

     below_med_income below_60pct_med_income below_moderate_income  \
0                 Yes                    Yes                   Yes   
1                 Yes                    Yes                   Yes   
2                 Yes                    Yes                   Yes   
3                 Yes                    Yes                   Yes   
4                 Yes                     No                   Yes   
...               ...                    ...                   ...   
2490               No                     No                    No   
2491               No                     No                    No   
2492               No                     No                    No   
2493               No                     No                    No   
2494               No                     No                   Yes   

        sup_dist                 csa            spa   Shape__Area  \
0     District 1  Wholesale District  SPA 4 - Metro  1.041050e+07   
1     District 1     Lincoln Heights  SPA 4 - Metro  3.724107e+06   
2     District 1     Lincoln Heights  SPA 4 - Metro  3.296129e+06   
3     District 1     Lincoln Heights  SPA 4 - Metro  4.782361e+06   
4     District 1           El Sereno  SPA 4 - Metro  1.099246e+07   
...          ...                 ...            ...           ...   
2490  District 4            Cerritos   SPA 7 - East  1.555650e+07   
2491  District 4            Cerritos   SPA 7 - East  1.421767e+07   
2492  District 4            Cerritos   SPA 7 - East  1.938903e+07   
2493  District 4            Cerritos   SPA 7 - East  1.866694e+07   
2494  District 4            Cerritos   SPA 7 - East  5.352220e+07   

      Shape__Length  
0      13808.463240  
1       9459.391827  
2       8868.744225  
3      10141.728020  
4      15893.383640  
...             ...  
2490   21274.227410  
2491   15905.089170  
2492   21218.412990  
2493   19500.866810  
2494   37625.839750  

[2458 rows x 11 columns]